src/components/ConfigureSupersetEmbeddedDashboardModal/UpdateSupersetEmbeddedDashboard.js

```diff
@@ -0,0 +1,219 @@
+}
+
+const parseErrorText = (error) =>
+    error?.details?.response?.errorReports[0]?.message ??
+    i18n.t('An unknown error occurred')
+
+export const UpdateSupersetEmbeddedDashboard = ({ closeModal }) => {
+    const dispatch = useDispatch()
+    const history = useHistory()
+    const id = useSelector(sGetSelectedId)
+    const {
+        loading: queryLoading,
+        error: queryError,
+        data: queryData,
+    } = useDataQuery(getDashboardQuery, { variables: { id } })
+    const dashboard = queryData?.dashboard
+    const name = dashboard?.name
+    const [mutationLoading, setMutationLoading] = useState(false)
+    const handeMutationError = useCallback(() => setMutationLoading(false), [])
+    const [updateDashboard, { error: updateError }] = useDataMutation(
+        updateDashboardQuery,
+        { variables: { id }, onError: handeMutationError }
+    )
+    const [deleteDashboard, { error: deleteError }] = useDataMutation(
+        deleteDashboardQuery,
+        { variables: { id }, onError: handeMutationError }
+    )
+    const mutationError = updateError || deleteError
+    const mutationErrorTitle = updateError
+        ? i18n.t('Could not update dashboard {{name}}', { name })
+        : i18n.t('Could not delete dashboard {{name}}', { name })
+    const mutationErrorText = parseErrorText(mutationError)
+    const {
+        hasFieldChanges,
+        isSupersetEmbedIdValid,
+        isSupersetEmbedIdFieldTouched,
+        values,
+        onChange,
+        onSupersetEmbedIdFieldBlur,
+        resetStateWithNewValues,
+    } = useSupersetEmbeddedDashboardFieldsState()
+    const handleSubmit = useCallback(
+        async (event) => {
+            event.preventDefault()
+            setMutationLoading(true)
+            await updateDashboard({ values })
+            await dispatch(tSetSelectedDashboardById(id))
+            setMutationLoading(false)
+            closeModal()
+        },
+        [values, updateDashboard, closeModal, dispatch, id]
+    )
+    const handleDelete = useCallback(
+        async (_, event) => {
+            event.preventDefault()
+            setMutationLoading(true)
+            await deleteDashboard()
+            dispatch(acClearSelected())
+            await dispatch(tFetchDashboards())
+            setMutationLoading(false)
+            closeModal()
+            history.push('/')
+        },
+        [deleteDashboard, closeModal, dispatch, history]
+    )
+
+    useEffect(() => {
+        if (dashboard) {
+            resetStateWithNewValues({
+                title: dashboard.name,
+                code: dashboard.code,
+                description: dashboard.description,
+                supersetEmbedId: dashboard.embedded.id,
+                showChartControls:
+                    !dashboard.embedded.options.hideChartControls,
+                showFilters: dashboard.embedded.options.filters.visible,
+            })
+        }
+    }, [dashboard, resetStateWithNewValues])
```

Comment:
Because I need to fetch the dashboard first, I ended up adding a new method to the `useSupersetEmbeddedDashboardFieldsState` hook. 

I considered using a nested component instead, which would load the dashboard in the parent and show the fields in the child (which could then be populated at mount time). 

The reason I went with this option is because this allows us to show the fields when the modal opens and thus will prevent the modal height from changing. But in the end this file turned out to be over 200 lines long, so perhaps splitting things up would have been a better option.

Let me know what you think.

Reply:
I decided to do different type of split. I've kept all the JSX in this component but have extracted the mutation logic into a hook. Quite happy with this new implementation.